packages/medusa/src/services/payment-collection.ts

```diff
@@ -160,4 +253,479 @@ export default class PaymentCollectionService extends TransactionBaseService {
+      })
+
+      if (payCol.status !== PaymentCollectionStatus.NOT_PAID) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_ALLOWED,
+          `Cannot set payment sessions for a payment collection with status ${payCol.status}`
+        )
+      }
+
+      sessionsInput = sessionsInput.filter((session) => {
+        return !!payCol.region.payment_providers.find(({ id }) => {
+          return id === session.provider_id
+        })
+      })
+
+      if (!this.isValidTotalAmount(payCol.amount, sessionsInput)) {
+        throw new MedusaError(
+          MedusaError.Types.UNEXPECTED_STATE,
+          `The total amount must be equal to ${payCol.amount}`
+        )
+      }
+
+      let customer: Customer | undefined = undefined
+
+      const selectedSessionIds: string[] = []
+      const paymentSessions: PaymentSession[] = []
+
+      for (const session of sessionsInput) {
+        if (!customer) {
+          customer = await this.customerService_.retrieve(session.customer_id, {
+            select: ["id", "email", "metadata"],
+          })
+          if (!customer) {
+            throw new MedusaError(
+              MedusaError.Types.UNEXPECTED_STATE,
+              `Invalid customer_id`
+            )
+          }
+        }
+
+        const existingSession = payCol.payment_sessions?.find(
+          (sess) => session.session_id === sess?.id
+        )
+
+        const inputData: PaymentProviderDataInput = {
+          resource_id: payCol.id,
+          currency_code: payCol.currency_code,
+          amount: session.amount,
+          provider_id: session.provider_id,
+          customer,
+          metadata: {
+            resource_id: payCol.id,
+          },
+        }
+
+        if (existingSession) {
+          const paymentSession = await this.paymentProviderService_
+            .withTransaction(manager)
+            .updateSessionNew(existingSession, inputData)
+
+          selectedSessionIds.push(existingSession.id)
+          paymentSessions.push(paymentSession)
+        } else {
+          const paymentSession = await this.paymentProviderService_
+            .withTransaction(manager)
+            .createSessionNew(inputData)
+
+          selectedSessionIds.push(paymentSession.id)
+          paymentSessions.push(paymentSession)
+        }
+      }
+
+      if (payCol.payment_sessions?.length) {
+        const removeIds: string[] = payCol.payment_sessions
+          .map((sess) => sess.id)
+          .filter((id) => !selectedSessionIds.includes(id))
+
+        if (removeIds.length) {
+          await paymentCollectionRepository
+            .createQueryBuilder()
+            .delete()
+            .from(PaymentSession)
+            .where("id IN (:...ids)", { ids: removeIds })
+            .execute()
+        }
+      }
+
+      payCol.payment_sessions = paymentSessions
+      await paymentCollectionRepository.save(payCol)
+
+      return payCol
```

Comment:
nit, suggestion(non-blocking):
```suggestion
      return await paymentCollectionRepository.save(payCol)
```